llpc/lower/llpcSpirvLowerGlobal.cpp

```diff
@@ -370,163 +379,221 @@ static bool hasVertexIdx(const Constant &metaVal) {
+  auto handle = [&](const unsigned addrSpace) {
+    if (addrSpace != SPIRAS_Input && addrSpace != SPIRAS_Output)
+      return false;
+    // Skip if "load" instructions are not expected to be handled
+    const bool isTcsInput = (m_shaderStage == ShaderStageTessControl && addrSpace == SPIRAS_Input);
+    const bool isTcsOutput = (m_shaderStage == ShaderStageTessControl && addrSpace == SPIRAS_Output);
+    const bool isTesInput = (m_shaderStage == ShaderStageTessEval && addrSpace == SPIRAS_Input);
+    return isTcsInput || isTcsOutput || isTesInput;
+  };
 
-      const unsigned elemCount = inOutTy->getArrayNumElements();
-      for (unsigned i = 0; i < elemCount; ++i) {
-        Value *vertexIdx = ConstantInt::get(Type::getInt32Ty(*m_context), i);
-        auto elemValue = addCallInstForInOutImport(elemTy, addrSpace, elemMeta, nullptr, 0, nullptr, vertexIdx,
-                                                   InterpLocUnknown, nullptr, &loadInst);
-        loadValue = InsertValueInst::Create(loadValue, elemValue, {i}, "", &loadInst);
+  for (GlobalVariable &global : m_module->globals())
+    for (User *user : global.users()) {
+      if (LoadInst *loadInst = dyn_cast<LoadInst>(user)) {
+        // The user is a load
+        const unsigned addrSpace = global.getType()->getPointerAddressSpace();
+        if (!handle(addrSpace))
+          continue;
+        handleLoadInstGlobal(*loadInst, addrSpace);
+      } else if (GetElementPtrInst *gep = dyn_cast<GetElementPtrInst>(user)) {
+        // The user is a GEP
+        const unsigned addrSpace = gep->getType()->getPointerAddressSpace();
+        if (!handle(addrSpace))
+          continue;
+        // We look for load instructions in the GEP users
+        for (User *gepUser : gep->users()) {
+          // We shouldn't have any chained GEPs here, they are coalesced by the LowerAccessChain pass.
+          assert(!dyn_cast<GetElementPtrInst>(gepUser));
+          if (LoadInst *loadInst = dyn_cast<LoadInst>(gepUser))
+            handleLoadInstGEP(gep, *loadInst, addrSpace);
+        }
       }
-    } else {
-      loadValue = addCallInstForInOutImport(inOutTy, addrSpace, inOutMetaVal, nullptr, 0, nullptr, nullptr,
-                                            InterpLocUnknown, nullptr, &loadInst);
     }
-  }
-  m_loadInsts.insert(&loadInst);
-  loadInst.replaceAllUsesWith(loadValue);
 }
 
 // =====================================================================================================================
-// Visits "store" instruction.
+// Handle a single "store" instruction directly storing a global.
 //
-// @param storeInst : "Store" instruction
-void SpirvLowerGlobal::visitStoreInst(StoreInst &storeInst) {
-  Value *storeValue = storeInst.getOperand(0);
+// @param storeInst : Store instruction
+void SpirvLowerGlobal::handleStoreInstGlobal(StoreInst &storeInst) {
   Value *storeDest = storeInst.getOperand(1);
+  Value *storeValue = storeInst.getOperand(0);
```

Comment:
There is `storeInst.getPointerOperand()` and `storeInst.getValueOperand()` for more descriptive names (but I think it’s also fine to use `getOperand`).